### PR TITLE
Introduce `Scalar` type for ColumnarValue

### PR DESCRIPTION
**Pull request by notfilippo**
This PR represents the first step originating from experiment #11978, which itself stems from the broad objective described in proposal #11513.

---

## Rationale for this change

This change introduces the knowledge of `Scalar` type which identifies the physical representation of a scalar value. `Scalar` will replace `ScalarValue` in the `ColumnarValue` enum. This is done in order to allow, in future PRs, to remove some redundant variants of `ScalarValue` and transition to logical representation of scalar values in datafusion.

## Design rationale

The new `Scalar` type embeds both a `ScalarValue` and a `DataType`. While the type information might seem redundant considering the current implementation `ScalarValue` it will be beneficial once we start to remove redundant variants (such as replacing `ScalarValue::Utf8View` and `ScalarValue::LargeUtf8` while keeping `ScalarValue::Utf8`) in order to keep track of the represented type during physical execution.

### Alternatives considered: [`arrow::array::Scalar`](https://docs.rs/arrow/latest/arrow/array/struct.Scalar.html)

Arrow's array crate provides a `Scalar` type which implements Datum and which holds a reference to a `len() = 1` arrow array. While this type fully identifies the physical representation of a scalar value it falls short when considering the overhead of copying a fully rendered ArrayRef instead of primitives types.

This possibility was also already discussed here: https://github.com/apache/datafusion/issues/7353#issuecomment-1739664804.

## What changes are included in this PR?

- Introduced the `Scalar` struct
- Replaced `ScalarValue` with `Scalar` in `ColumnarValue`'s Scalar variant
- Adapted existing code to the Scalar variant change

## Are these changes tested?

This change should be a no-op as it doesn't include any behavioural change. The existing test suite is sufficient to verify this change.

## TODO

- [ ] rustdoc on new types and functions

### CODE REVIEW

**Comment by notfilippo** [comment]
The main change is contained in this file: https://github.com/apache/datafusion/pull/12536/files#diff-2cc034babb8e7f8601dda34ecaa2119104eccd76d8ad2f9e19b26b01463634d6

All other files are mostly updates to support the change from `ScalarValue` to `Scalar` in ColumnarValue

**Comment by alamb** [comment]
FYI @findepi

**Comment by alamb** [comment]
I will try and find time to review this tomorrow

**Comment by findepi** [comment]
Thank you @notfilippo for working on this!

> `ScalarValue::Utf8View` and `ScalarValue::LargeUtf8` while keeping `ScalarValue::Utf8`) in order to keep track of the represented type during physical execution.

Can you elaborate what this is needed for?

**Comment by notfilippo** [comment]
> Can you elaborate what this is needed for?

@findepi -- `DataType` in Scalar is needed as a way to retain the physical type that the ScalarValues represents. That knowledge will be used when creating arrays from scalar types. It comes especially useful when converting a vector of scalars into an array of the correct physical type.

You can find some more context here: https://github.com/apache/datafusion/pull/11978#issuecomment-2293837444

As @jayzhan211 mentions in https://github.com/apache/datafusion/pull/11978#issuecomment-2294518747 and https://github.com/apache/datafusion/pull/11978#issuecomment-2297870220 it's a step towards the goal but it does not represent the full decoupling of physical and logical types.

**Review by jayzhan211 — DISMISSED** [review]
I'm fine with this change. Let's wait for others' views

**Comment by notfilippo** [comment]
Before merging I was also planning to add some rustdoc on the new type in order to explain its purpose and future plans.

File: datafusion/expr-common/src/scalar.rs
Comment (findepi) [inline datafusion/expr-common/src/scalar.rs:27]: Should I assume that constructing `Scalar { value:  ScalarValue::Int32(..), data_type: DataType::Null }` is not valid?
We need to define constraints on these values and ideally enforce them.

File: datafusion/expr-common/src/scalar.rs
Comment (findepi) [inline datafusion/expr-common/src/scalar.rs:34]: This code shows that Scalar is a redundant wrapper around ScalarValue.
I understand this is transitionary. Can we capture the desired move forward in the code?
It's great if the code self-documents, but without any commentary this code self-documents itself as being redundant

File: datafusion/expr-common/src/scalar.rs
Comment (findepi) [inline datafusion/expr-common/src/scalar.rs:47]: Conversion from a data type to a value should be less implicit.
Can we call this function `new_null_of(type)`?

File: datafusion/expr-common/src/columnar_value.rs
Comment (findepi) [inline datafusion/expr-common/src/columnar_value.rs:94]: when i read @jayzhan211 's https://github.com/apache/datafusion/pull/12488#issuecomment-2359844558 o thought that ScalarValue is going to be left in ColumnarValue (least overhead, as relating to @alamb 's https://github.com/apache/datafusion/pull/12488#pullrequestreview-2312833258).
But here is't being replaced.

What's the next step here? what's the end state?

File: datafusion/expr-common/src/columnar_value.rs
Comment (findepi) [inline datafusion/expr-common/src/columnar_value.rs:201]: Why? Don't we need to transform the actual value, eg from int to string or vice versa?

**Comment by findepi** [comment]
Brief code-level documentation -- what this new Scalar is, what it is not, what are the constraints and what's the goal of this evolution is a must for this.

I would like to review again once this is added, so that I understand better what we're trying to accomplish.

The referenced issue https://github.com/apache/datafusion/issues/11513 is a good record of all the discussions, but it's not clear which comments are opinions that were later refined and which are normative. 

Thanks for the patience.

File: datafusion/expr-common/src/scalar.rs
Comment (notfilippo) [reply datafusion/expr-common/src/scalar.rs:47]: I agree that the conversion should look less implicit so I'll move it to a dedicated function. I should mention that `ScalarValue` uses the same pattern:

https://github.com/apache/datafusion/blob/21ec3328428d4f27aa84abf7bd0c58332edcaed5/datafusion/common/src/scalar/mod.rs#L3354-L3361

and I adopted it to be consistent with that API.

File: datafusion/expr-common/src/scalar.rs
Comment (notfilippo) [reply datafusion/expr-common/src/scalar.rs:27]: I think those constraint can be made implicit if we consider editing the possible way the `Scalar` object can be constructed. By only keeping the `::from(ScalarValue)` and the `::try_from_array(...)` method (which is used for casting) we artificially limit the possible values that `data_type` can hold to the ones that are cast-compatible with `value`.

File: datafusion/expr-common/src/scalar.rs
Comment (notfilippo) [reply datafusion/expr-common/src/scalar.rs:34]: Totally understandable. As I mentioned https://github.com/apache/datafusion/pull/12536#issuecomment-2363690150 I wanted to first gauge opinions about my implementation (which proved successful as you raised a lot of interesting points to account for) before committing on some documentation prose.

File: datafusion/expr-common/src/columnar_value.rs
Comment (notfilippo) [reply datafusion/expr-common/src/columnar_value.rs:201]: Leftover from wrong notes that I was taking while editing the file. Removed.

File: datafusion/expr-common/src/columnar_value.rs
Comment (notfilippo) [reply datafusion/expr-common/src/columnar_value.rs:94]: The next step is to slowly remove variants of `ScalarValue` while still accounting for them via `data_type` (which in my opinion is the least intrusive change to support this effort). After we are satisfied with the variants that remain we reconsider the logic in expressions and operators in order to support what @jayzhan211 in proposing:

> I think we could have LogicalType without any Arrow's DataType contained in it in the future

by sourcing the `data_type` directly from the execution schema.

File: datafusion/expr-common/src/scalar.rs
Comment (alamb) [reply datafusion/expr-common/src/scalar.rs:34]: As I understood the idea, eventually the DataType of the scalar will be different than the underlying representation

I 100% agree with the sentiment that this intent should be captured in code comments as it is very much non obvious from the code

File: datafusion/functions/src/encoding/inner.rs
Comment (alamb) [inline datafusion/functions/src/encoding/inner.rs:269]: The switch from explicit use of a variant to a `From` clause is a nice change regardless, FWIW, and we could potentially break it out into its own PR

File: datafusion/expr-common/src/scalar.rs
Comment (alamb) [inline datafusion/expr-common/src/scalar.rs:72]: We could potentially call this function `as_value` to hit that it was returning a reference and was cheap

File: datafusion/expr-common/src/scalar.rs
Comment (alamb) [reply datafusion/expr-common/src/scalar.rs:27]: Another potential way to represent this idea would be to use an enum, perhaps, like

```rust
pub enum Scalar {
  /// A ScalarValue (transitory)
  ScalarValue(ScalarValue),
  /// A logical String with associated Arrow physical typ
  String {
    value: Option<String>,
    data_type: DataType,
  },
}
```

I think we could modify this in the future as well

**Review by alamb — COMMENTED** [review]
First of all thank you @notfilippo and @findepi  and @jayzhan211 -- I think this is a very nice step towards splitting out logical and physical types. 

I took a brief look through this PR and it seems pretty good to me (other than lack of comments)

So one way to proceed if we are agreed on the basic idea is to create a new "Epic" ticket that lists the steps planned (it is fine if we don't know them all now). Here is an example https://github.com/apache/arrow-rs/issues/5374

The first few steps might be
- [ ] Add `Scalar`
- [ ] Remove `ScalarValue::LargeString` in favor of `Scalar`
- [ ] ...

Having such an epic would also make it clear what the current plan was (as opposed to the various options that had been previously discussed)

cc @ozankabak  as this project appears to be moving along

File: datafusion/expr-common/src/scalar.rs
Comment (findepi) [reply datafusion/expr-common/src/scalar.rs:27]: @alamb  what would be "a logical String with associated data type"? i.e. what value of target data type does it represent and how to obtain it? And it it costly to obtain it during optimization process?
i didn't think about this too deeply yet, but would it be possible for the logical IR to avoid ambiguities?

> I think those constraint can be made implicit if we consider editing the possible way the `Scalar` object can be constructed. By only keeping the `::from(ScalarValue)` and the `::try_from_array(...)` method (

@notfilippo that's great there are constraints. 
let's make sure they are documented -- would be useful for the readers, and also would capture the design intent in the code.

**Comment by ozankabak** [comment]
Will carefully review next week. Thank you for working on this

File: datafusion/expr-common/src/scalar.rs
Comment (findepi) [reply datafusion/expr-common/src/scalar.rs:34]: > before committing on some documentation prose.

that's a good point
and actually, i have hard time with long documentation prose -- i am maybe a slow reader.
what i am looking for is exactly what you need -- to gauge opinions you need to understand what the code says today, but also what (today) it is supposed to say tomorrow

> As I understood the idea, eventually the DataType of the scalar will be different than the underlying representation

@alamb , on the face value it seems to be opposite to what i understood from @notfilippo 's https://github.com/apache/datafusion/pull/12536#discussion_r1769524231. I don't have strong opinion yet what scalar should or should not be, but would be great to understand the desired end state. Can you please help me with that?

File: datafusion/expr-common/src/scalar.rs
Comment (findepi) [reply datafusion/expr-common/src/scalar.rs]: > I should mention that `ScalarValue` uses the same pattern:

yeah, i know. i still couldn't resist :)

File: datafusion/expr-common/src/columnar_value.rs
Comment (findepi) [reply datafusion/expr-common/src/columnar_value.rs:94]: - Is ScalarValue supposed to be in the logical types layer, or physical?
- is Scalar supposed to be in the logical types layer, or physical?
- is ColumnarValue purely physical layer?

> remove variants of `ScalarValue` while still accounting for them via `data_type`

what do we need this for?
(maybe that's obvious for someone with clarity on the preceding questions)

File: datafusion/expr-common/src/scalar.rs
Comment (findepi) [reply datafusion/expr-common/src/scalar.rs:72]: eg arrow `Field::name()` is named just "name". isn't `as_foo` more appropriate for representhing self as foo, rather than reading a foo attribute of self?

**Comment by findepi** [comment]
> So one way to proceed if we are agreed on the basic idea is to create a new "Epic" ticket that lists the steps planned (it is fine if we don't know them all now). Here is an example [apache/arrow-rs#5374](https://github.com/apache/arrow-rs/issues/5374)

I love the idea, @alamb !

just note that only maintainers can add to such epic.

> Having such an epic would also make it clear what the current plan was (as opposed to the various options that had been previously discussed)

that would be helpful indeed!

File: datafusion/expr-common/src/columnar_value.rs
Comment (jayzhan211) [reply datafusion/expr-common/src/columnar_value.rs:94]: I think it is hard to tell the role given we are in the intermediate state

The end state in my mind is that,
`ColumnarValue` stores `ArrayRef` for multiple rows and `ScalarValue` for single rows case.
We will have `LogicalType` which could be resolved from `DataType` given the mapping we determined, which should be customizable as well for user defined type. Similar to the role `Scalar` in this PR, which record the relationship between `ScalarValue` to `DataType`.

Comment (jayzhan211) [reply datafusion/expr-common/src/columnar_value.rs:94]: > remove variants of ScalarValue while still accounting for them via data_type


Our first step is to minimize the variant of types, so we don't need `ScalarValue::Utf8`, `ScalarValue::LargeUtf8` any more but `Scalar` which has `ScalarValue::String` + `DataType::Utf8` or `ScalarValue::String` + `DataType::LargeUtf8`

File: datafusion/expr-common/src/scalar.rs
Comment (notfilippo) [reply datafusion/expr-common/src/scalar.rs:34]: > > As I understood the idea, eventually the DataType of the scalar will be different than the underlying representation
> 
> @alamb , on the face value it seems to be opposite to what i understood from @notfilippo 's [#12536 (comment)](https://github.com/apache/datafusion/pull/12536#discussion_r1769524231). I don't have strong opinion yet what scalar should or should not be, but would be great to understand the desired end state. Can you please help me with that?

The main point of this PR is to remove variants of ScalarValue and represent them instead through the DataType contained in the Scalar type (which is what @\alamb is saying). The constraints put in place by the current constructors of Scalar allow it to only contain DataType fully compatible with the ScalarType provided (see `::from(ScalarValue)`) or cast-compatible (see `::try_from_array(...)`). Will try to make it clear via docs later today.

File: datafusion/expr-common/src/scalar.rs
Comment (notfilippo) [reply datafusion/expr-common/src/scalar.rs]: Sounds good! Updated 😄

File: datafusion/expr-common/src/columnar_value.rs
Comment (findepi) [reply datafusion/expr-common/src/columnar_value.rs:94]: > The end state in my mind is that,
> `ColumnarValue` stores `ArrayRef` for multiple rows and `ScalarValue` for single rows case. `ScalarValue` has either native type like `i64` or `arrow::Scalar`.

Ok, so `ScalarValue` is going to be physical and also going to be simplified.

what will be used in `Expr::Literal`?
i am very interested in this part


> `ScalarValue::String` is then something close to `logical type`

but being used in very physical context, right?

Comment (jayzhan211) [reply datafusion/expr-common/src/columnar_value.rs:94]: For value, there are `ScalarValue` and `ArrayRef`, for type there are `LogicalType` and `DataType`.

`Scalar` will be used in `Expr::Literal`, it contains `ScalarValue::String` + `DataType`. We can get the logical type from `DataType` or `ScalarValue` iteslf in the logical layer. And, decode the `String` with `DataType` in the physical layer.

Comment (jayzhan211) [reply datafusion/expr-common/src/columnar_value.rs:94]: For `ScalarValue::List`, we might need something like `ScalarValue::List(Vec<ScalarValue>)` instead of `arrow::Scalar` 🤔

Comment (jayzhan211) [reply datafusion/expr-common/src/columnar_value.rs:94]: `ScalarValue` should be used all over datafusion, not restricted to logical type or physical type. We just need to bring along with `DataType` so we understand how to decode the value.

**Comment by notfilippo** [comment]
> So one way to proceed if we are agreed on the basic idea is to create a new "Epic" ticket that lists the steps planned (it is fine if we don't know them all now). Here is an example [apache/arrow-rs#5374](https://github.com/apache/arrow-rs/issues/5374)

I love the idea! My starting suggestion for potential issues is:
- Add `Scalar`
- Use `Scalar` in `Expr::Literal`
- Introduce `LogicalType` enum and the _logically equivalent_ notion
- Remove `ScalarValue::LargeUtf8` and `ScalarValue::Utf8View` in favour of `ScalarValue::Utf8`
- Remove `ScalarValue::LargeBinary` and `ScalarValue::BinaryView` in favour of `ScalarValue::Utf8`
- Remove `ScalarValue::Dictionary` (from #12488)

**Comment by notfilippo** [comment]
Also, @findepi -- I've finally settled on some docs. 😄

**Comment by alamb** [comment]
> just note that only maintainers can add to such epic.

I think the author of a ticket can do so too. 

Another way that has worked in the past is just to leave comments and I or another maintainer can copy the links to the description

We could also explore adding people who are interested as "triage"ers (access to tickets/Prs without write access to code): https://github.com/apache/infrastructure-asfyaml?tab=readme-ov-file#triage

File: datafusion/expr-common/src/scalar.rs
Comment (alamb) [inline datafusion/expr-common/src/scalar.rs:35]: I don't think this is intentional (that some types can not be handled)

**Review by alamb — COMMENTED** [review]
Thanks -- this is looking good to me

Let's let @ozankabak  have a chance to look at it and then figure out how to move forward

Even if we just get to the point that ScalarValue has a different representation I think it would be an improvement in readability, though maybe that doesn't justify the code churn.

Maybe we can set up a zoom meeting for anyone interested in this topic to discuss and get alignment if possible (with a writeup afterwards) as suggested by @findepi

File: datafusion/expr-common/src/scalar.rs
Comment (notfilippo) [reply datafusion/expr-common/src/scalar.rs:35]: I'm not sure what you exactly mean. Currently (but potentially not in the future) the `DataType` of `Scalar` should be at least logically equivalent to the data type reported by `ScalarValue::data_type()`

**Comment by ozankabak** [comment]
This looks good to me. The main challenge I had was extrapolating/imagining what the post-transitionary state of things will really look like. At this point part of the code seem redundant (as in @findepi's inline comments).

I am unclear on what is the best way to proceed. I fear we may run into unforeseen issues (in downstream projects and/or upstream DF) as we take the subsequent steps (i.e. evolve `ScalarValue` and remove certain data types from it).

One way we can proceed is to create a branch and make progress on it and rebase/merge main weekly. We can discover upstream DF issues this way without disrupting the progress on `main`. Also, downstream projects can experiment with their cargo setup to point to this branch and see what happens. This would give us the opportunity to discover issues without causing disruptions and ensure a smooth transition into the new paradigm if things still look good as we go through the motions. Obviously this will introduce some development burden, but it will introduce unintended bad impact, reverts etc.

What do you all think about the path forward?

**Comment by notfilippo** [comment]
> One way we can proceed is to create a branch and make progress on it and rebase/merge main weekly. We can discover upstream DF issues this way without disrupting the progress on `main`. Also, downstream projects can experiment with their cargo setup to point to this branch and see what happens.

I love this approach. I think having this + the [Epic] tracking ticket will be a great way of managing progress for the proposal.

File: datafusion/expr-common/src/columnar_value.rs
Comment (findepi) [reply datafusion/expr-common/src/columnar_value.rs:94]: > , it contains `ScalarValue::String` + `DataType`

so a number is stored as "5"?
how do i compare two float scalar values? i need to parse them?

how do i represent a constant (aka "literal") of type `array<row<a: map<string, array<row<a: int, b: int>>>>>`?

i do believe constants in the plan are better expressed in their native representation, like we did in `ScalarValue` (or via `Any`, as long as we have type information needed to cast it back)

i guess this is a similar concern to one raised by @alamb here: https://github.com/apache/datafusion/pull/12536#discussion_r1774086062

Comment (jayzhan211) [reply datafusion/expr-common/src/columnar_value.rs:94]: Number is not strictly stored as string. ScalarValue::I32(5) is numeric, ScalarValue::String("5") is numeric string. ScalarValue::Float(1.1) is numeric, you can compare them like rust native float type.

Comment (jayzhan211) [reply datafusion/expr-common/src/columnar_value.rs:94]: I think there is nothing change for literal case, we still get `Literal(ScalarValue)`.

Comment (jayzhan211) [reply datafusion/expr-common/src/columnar_value.rs:94]: > `ScalarValue` should be used all over datafusion, it is more a **value** and has no logical or physical concept. We just need to bring along with `DataType` so we understand how to decode the value in physical layer and understand what logical type it is to interact within the logical plan

And bring along `DataType` could be a transition state, I guess we could get the DataType from schema so we don't even require `DataType` in `ScalarValue`

Comment (jayzhan211) [reply datafusion/expr-common/src/columnar_value.rs:94]: As long as we have `Schema` we can get the `DataType`, and we can get the LogicalType from it if each DataType has at most one LogicalType. `ScalarValue` is just a wrapper for the scalar case of `ArrayRef` or rust native type. We can carry `String` for `Utf8` / `Utf8View` / `LargeUtf8` and decode it if needed, and also carry `Scalar<ArrayRef>` for `List` / `LargeList` and so on.

Comment (findepi) [reply datafusion/expr-common/src/columnar_value.rs:94]: > we can get the LogicalType from it if each DataType has at most one LogicalType

We should not assume this to be the case.
I suppose https://github.com/apache/datafusion/issues/12644 may convey why, but even if we don't do Extension Types, this assumption will be very limiting.

Comment (jayzhan211) [reply datafusion/expr-common/src/columnar_value.rs:94]: > > we can get the LogicalType from it if each DataType has at most one LogicalType
> 
> We should not assume this to be the case. I suppose #12644 may convey why, but even if we don't do Extension Types, this assumption will be very limiting.

I'm still not convinced why my assumption is limited. For extension types, we could customize/extend the **logical type** for each `DataType`. It doesn't make sense to have more than 1 logical type that decoded to the same physical type, they are *equivalent* thus should be the same logical type. Even if they have historical issue thus require more than one type for backward compatibility, they are able to do it if they built their type on top of Datafusion's Logical Type.

**Comment by notfilippo** [comment]
So, should we progress with the approach @ozankabak suggested? cc @alamb @jayzhan211

**Comment by jayzhan211** [comment]
> So, should we progress with the approach @ozankabak suggested? cc @alamb @jayzhan211

Sure, we should work on the branch until the end state is clear to us.

**Comment by notfilippo** [comment]
> Sure, we should work on the branch until the end state is clear to us.

Sounds good 🚀 So should I switch the target of this PR to the new branch? Can a maintainer create this branch?

**Comment by alamb** [comment]
> > Sure, we should work on the branch until the end state is clear to us.
> 
> Sounds good 🚀 So should I switch the target of this PR to the new branch? Can a maintainer create this branch?

I have created the branch https://github.com/apache/datafusion/tree/logical-types

**Comment by notfilippo** [comment]
I've changed the base branch of this PR 👍

**Comment by alamb** [comment]
Let's merge to the branch and keep iterating there!

**Pull request merged**
